### PR TITLE
Support on_replace: :update for has_one/belongs_to

### PR DESCRIPTION
**Pull request by eniolopes**
Addresses https://github.com/elixir-ecto/ecto/issues/1627

### CODE REVIEW

**Comment by eniolopes** [comment]
@josevalim as spoken yesterday, I'm no entirely sure where to go now, need some guidance.

File: lib/ecto/changeset/relation.ex [hunk 1]
```diff
@@ -244,7 +248,7 @@ defmodule Ecto.Changeset.Relation do
       single_change(new, current, fun, [:update, :delete], true)
     else
       case on_replace(relation, current) do
-        {:ok, _} -> single_change(new, nil, fun, [:insert], false)
+        {:ok, changeset} -> single_change(new, nil, fun, [changeset.action], false)
```
Comment (josevalim) [inline lib/ecto/changeset/relation.ex]: If we are on a update, we need to call something similar to the line 248 above, passing both `new` and `current`, to ensure the new fields will be merged on top of current. Maybe changing [this line](https://github.com/eniolopes/ecto/blob/469f1f90f1647e70c823741c572f10e869ac1faa/lib/ecto/changeset/relation.ex#L247) to `if new_pks.(new) == current_pks.(current) or relation.on_change == :update do` is a better way to move forward?

**Comment by michalmuskala** [comment]
@eniolopes ping

**Comment by eniolopes** [comment]
@michalmuskala I've done the change requested by José, I think we have a working solution for `belongs_to`. Starting the work on `has_one` now.

**Comment by eniolopes** [comment]
@josevalim @michalmuskala I think I'm done with this, sorry for the delay. The tests I added will be enough? Let me know if there is anything else left to do here.

File: lib/ecto/changeset/relation.ex [hunk 2]
```diff
@@ -240,11 +240,11 @@ defmodule Ecto.Changeset.Relation do
   end
 
   defp single_change(relation, new, current_pks, new_pks, fun, current) do
-    if new_pks.(new) == current_pks.(current) do
+    if new_pks.(new) == current_pks.(current) or relation.on_replace == :update do
```
Comment (josevalim) [inline lib/ecto/changeset/relation.ex]: Let's pattern match `%{on_replace: on_replace} = relation` on the signature instead. :)

File: lib/ecto/association.ex [hunk 3]
```diff
@@ -343,15 +344,21 @@ defmodule Ecto.Association.Has do
...
-
     unless on_delete in @on_delete_opts do
       raise ArgumentError, "invalid :on_delete option for #{inspect name}. " <>
         "The only valid options are: " <>
         Enum.map_join(@on_delete_opts, ", ", &"`#{inspect &1}`")
     end
 
-    unless on_replace in @on_replace_opts do
+    on_replace = Keyword.get(opts, :on_replace, :raise)
+    cardinality = Keyword.get(opts, :cardinality, :raise)
```
Comment (josevalim) [inline lib/ecto/association.ex]: `Keyword.fetch!`. I believe the cardinality is always present.

Comment (josevalim) [inline lib/ecto/association.ex]: We should also change this line: https://github.com/eniolopes/ecto/blob/d844662db4c6cd33769ba4a1b5d4174fd02bed5b/lib/ecto/association.ex#L369

It now can use the `:cardinality` variable.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**